src/shogun/machine/GeneralizedLinearMachine.h

```diff
@@ -0,0 +1,215 @@
+ *
+ *	.. math::
+ *
+ *		\\mathcal{P}_2 = \\|\\Gamma \\beta \\|_2^2 \\
+ *		\\mathcal{P}_1 = \\sum_g \\|\\beta_{j,g}\\|_2
+ *
+ *	where :math:`\\Gamma` is the Tikhonov matrix: a square factorization
+ *	of the inverse covariance matrix and :math:`\\beta_{j,g}` is the
+ *	:math:`j` th coefficient of group :math:`g`.
+ *
+ *	The generalized L2 penalty defaults to the ridge penalty when
+ *	:math:`\\Gamma` is identity.
+ *
+ *	The generalized L1 penalty defaults to the lasso penalty when each
+ *	:math:`\\beta` belongs to its own group.
+ *
+ * */
+class GeneralizedLinearMachine : public Machine
+{
+ 	public:
+ 		/** default constructor */
+ 		GeneralizedLinearMachine();
+
+ 		/** destructor */
+ 		virtual ~GeneralizedLinearMachine();
+
+ 		/** copy constructor */
+ 		GeneralizedLinearMachine(const std::shared_ptr<GeneralizedLinearMachine>& machine);
+
+		/** get w
+		 *
+		 * @return weight vector
+		 */
+		virtual SGVector<float64_t> get_w() const;
+
+		/** set w
+		 *
+		 * @param src_w new w
+		 */
+		virtual void set_w(const SGVector<float64_t> src_w);
+
+		/** set bias
+		 *
+		 * @param b new bias
+		 */
+		virtual void set_bias(float64_t b);
+
+		/** get bias
+		 *
+		 * @return bias
+		 */
+		virtual float64_t get_bias() const;
+
+ 		/** set features
+ 		 *
+ 		 * @param feat features to set
+ 		 */
+ 		virtual void set_features(std::shared_ptr<DotFeatures> feat);
+
+		/**
+		* A specialization of the train_machine method
+		* @param feats training data
+		*/
+		virtual bool train_machine(std::shared_ptr<const DenseFeatures<float64_t>> feats);
+
+		/** apply linear machine to data
+		 * for regression problem
+		 *
+		 * @param data (test)data to be classified
+		 * @return classified labels
+		 */
+		virtual std::shared_ptr<RegressionLabels> apply_regression(std::shared_ptr<Features> data=NULL);
+
+ 		/** get features
+ 		 *
+ 		 * @return features
+ 		 */
+ 		virtual std::shared_ptr<DotFeatures> get_features();
+
+ 		/** Returns the name of the SGSerializable instance.  It MUST BE
+ 		 *  the CLASS NAME without the prefixed `C'.
+ 		 *
+ 		 * @return name of the SGSerializable
+ 		 */
+ 		virtual const char* get_name() const { return "GeneralizedLinearMachine"; }
+
+	protected:
+
+		/** apply get outputs
+		 *
+		 * @param data features to compute outputs
+		 * @return outputs
+		 */
+		virtual SGVector<float64_t> apply_get_outputs(std::shared_ptr<Features> data);
+
+		/**Conditional intensity function.*/
+		virtual SGVector<float64_t> conditional_intensity(SGMatrix<float64_t> X, SGVector<float64_t> w, float64_t bias);
+
+		/** predict for one vector */
+		virtual SGVector<float64_t> predict(SGMatrix<float64_t> X);
+
+		/** fit model */
+		virtual bool fit(SGMatrix<float64_t> X, SGVector<float64_t> y);
+
+		/** compute gradient of weights */
+		virtual SGVector<float64_t> compute_grad_L2_loss_w(SGMatrix<float64_t> X, SGVector<float64_t> y, SGVector<float64_t> w, float64_t bias);
```

Comment:
these methods here can of course say in terms of matrices. I think they should all have const references in the parameters though